Invoke Platform Code Works For One Platform Only, But Flagged As Wrong For The Two Others

### Description

First, a quick compliment to the entire .NET MAUI team! Having used Xamarin.Forms for several years, it's cool to see a new version with better structuring.

The issue I'm reporting here has to do with what Microsoft calls "invoking platform code", or (as I understand it) the replacement to the DependencyService object used in Xamarin.Forms. Simply put, the compiler appears to be fine with just one implementations of a base partial class, but not for the two others.

For a quick review of invoking platform code - and where I got the instructions to test - please see this page:

[](https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/invoke-platform-code)

This attached project demonstrates the nature of the issue. (The "bin" and "obj" folders were removed to reduce space.) It is based on the default project created when a new .NET MAUI one is selected.

[invoke-platform-bug-demo.zip](https://github.com/dotnet/maui/files/8250379/invoke-platform-bug-demo.zip)

In this package the following files are pertinent:

**DummyClass.cs, Shared/Root folder** - defines a namespace, a public partial class within that namespace, and a signature for a public partial method, DummyClassMethod().

**Platforms/Android/DummyClass.cs** - implementation of the method DummyClassMethod() with a single Debug.WriteLine statement indicating the Android platform is being called.

**Platforms/iOS/DummyClass.cs** - implementation of the method DummyClassMethod() with a single Debug.WriteLine statement indicating the iOS platform is being called.

**Platforms/Windows/DummyClass.cs** - implementation of the method DummyClassMethod() with a single Debug.WriteLine statement indicating the Windows platform is being called.

In two of the DummyClass.cs files, in the code editor, the compiler will flag the call to DummyClassMethod() with the following error message:

`"CS0757: A partial method may not have multiple implementing declarations."`

If I understand the Microsoft page above, then what I've coded should work. But in at least two cases, error CS0757 arises.

As I understand it, in order for this scheme to work, all of the following must be true:

- All files MUST refer to the same namespace.
- All classes MUST be public and use the same name, as well as marked "partial."

I've looked at one other blog and its information matches that of the Microsoft page, so I'm pretty sure my coding is correct. But if I missed something, I'd appreciate being apprised. 


### Steps to Reproduce

1. Create a new .NET MAUI project in Visual Studio.
2. In the shared project (root), create a new partial class under a new namespace. Add a signature for one partial method to the namespace without an implementation.
3. In each of the folders under their respective platforms - under "Platforms/" then the platform name - create a file of the same name, referencing the same namespace as in step 2, and the same partial class also, with a call to the partial method, but with an implementation of at least one executable line (e.g., a Debug.WriteLine statement). 
4. Look for the red squiggles under the method calls in at least two of the method calls for two of the platforms.

### Version with bug

Preview 13 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11, Windows 11, iOS uncertain

### Did you find any workaround?

The only workaround I'm tempted to use is to fall back to the DependencyService but I'd rather not do that. I'd like for this to be working. I'm attempting to learn the new platform.

### Relevant log output

_No response_

--- a/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
+++ b/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
@@ -71,38 +71,38 @@
   <!-- Add metadata indicating that the platform-specific files are not part of every build configuration. -->
   <ItemGroup Condition=" '$(SingleProject)' == 'true' ">
     <Compile Update="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <LimitedConfigurationFile>true</LimitedConfigurationFile>
+      <ExcludeFromCurrentConfiguration>true</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
   <!-- Add metadata for the files that are actually part of the current build configuration. -->
   <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'android' ">
     <Compile Update="$(AndroidProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+      <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
   <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'ios' ">
     <Compile Update="$(iOSProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+      <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
   <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'maccatalyst' ">
     <Compile Update="$(MacCatalystProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+      <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
   <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'windows' ">
     <Compile Update="$(WindowsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+      <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
   <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'tizen' ">
     <Compile Update="$(TizenProjectFolder)**/*$(DefaultLanguageSourceExtension)">
-      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+      <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>
 
@@ -119,7 +119,7 @@
     <ItemGroup>
       <!-- Remove everything that isn't part of this platform -->
       <Compile
-          Condition=" '%(Compile.LimitedConfigurationFile)' == 'true' and '%(Compile.IncludeInThisConfiguration)' != 'true' "
+          Condition=" '%(Compile.ExcludeFromCurrentConfiguration)' == 'true' "
           Remove="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
 
       <!-- Remove all Windows (WinUI) XAML Files from the Windows folder -->

--- a/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
+++ b/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
@@ -68,6 +68,44 @@
     <TizenSharedPrefix>$(TizenProjectFolder)shared</TizenSharedPrefix>
   </PropertyGroup>
 
+  <!-- Add metadata indicating that the platform-specific files are not part of every build configuration. -->
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' ">
+    <Compile Update="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <LimitedConfigurationFile>true</LimitedConfigurationFile>
+    </Compile>
+  </ItemGroup>
+
+  <!-- Add metadata for the files that are actually part of the current build configuration. -->
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'android' ">
+    <Compile Update="$(AndroidProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'ios' ">
+    <Compile Update="$(iOSProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'maccatalyst' ">
+    <Compile Update="$(MacCatalystProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'windows' ">
+    <Compile Update="$(WindowsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'tizen' ">
+    <Compile Update="$(TizenProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+      <IncludeInThisConfiguration>true</IncludeInThisConfiguration>
+    </Compile>
+  </ItemGroup>
+
   <!--
     Run before both _MauiInjectXamlCssAdditionalFiles and GenerateMSBuildEditorConfigFileShouldRun because
     if for some reason the _MauiInjectXamlCssAdditionalFiles target is not run, we still get in at the
@@ -79,23 +117,10 @@
 
     <!-- Removals -->
     <ItemGroup>
-      <Compile Remove="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
-      <!-- Explicitly include the platform code -->
+      <!-- Remove everything that isn't part of this platform -->
       <Compile
-          Condition=" '$(TargetPlatformIdentifier)' == 'android' and '$(AndroidProjectFolder)' != '' "
-          Include="$(AndroidProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
-      <Compile
-          Condition=" '$(TargetPlatformIdentifier)' == 'ios' and '$(iOSProjectFolder)' != '' "
-          Include="$(iOSProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
-      <Compile
-          Condition=" '$(TargetPlatformIdentifier)' == 'maccatalyst' and '$(MacCatalystProjectFolder)' != '' "
-          Include="$(MacCatalystProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
-      <Compile
-          Condition=" '$(TargetPlatformIdentifier)' == 'windows' and '$(WindowsProjectFolder)' != '' "
-          Include="$(WindowsProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
-      <Compile
-          Condition=" '$(TargetPlatformIdentifier)' == 'tizen' and '$(TizenProjectFolder)' != '' "
-          Include="$(TizenProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
+          Condition=" '%(Compile.LimitedConfigurationFile)' == 'true' and '%(Compile.IncludeInThisConfiguration)' != 'true' "
+          Remove="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)" />
 
       <!-- Remove all Windows (WinUI) XAML Files from the Windows folder -->
       <_MauiXamlToRemove

--- a/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
+++ b/.nuspec/Microsoft.Maui.Controls.SingleProject.targets
@@ -68,40 +68,36 @@
     <TizenSharedPrefix>$(TizenProjectFolder)shared</TizenSharedPrefix>
   </PropertyGroup>
 
-  <!-- Add metadata indicating that the platform-specific files are not part of every build configuration. -->
   <ItemGroup Condition=" '$(SingleProject)' == 'true' ">
+    <!-- Add metadata indicating that the platform-specific files are not part of every build configuration. -->
     <Compile Update="$(PlatformsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>true</ExcludeFromCurrentConfiguration>
     </Compile>
-  </ItemGroup>
 
-  <!-- Add metadata for the files that are actually part of the current build configuration. -->
-  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'android' ">
-    <Compile Update="$(AndroidProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+    <!-- Add metadata for the files that are actually part of the current build configuration. -->
+    <Compile
+        Condition=" '$(TargetPlatformIdentifier)' == 'android' "
+        Update="$(AndroidProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
-  </ItemGroup>
-
-  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'ios' ">
-    <Compile Update="$(iOSProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+    <Compile
+        Condition=" '$(TargetPlatformIdentifier)' == 'ios' "
+        Update="$(iOSProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
-  </ItemGroup>
-
-  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'maccatalyst' ">
-    <Compile Update="$(MacCatalystProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+    <Compile
+        Condition=" '$(TargetPlatformIdentifier)' == 'maccatalyst' "
+        Update="$(MacCatalystProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
-  </ItemGroup>
-
-  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'windows' ">
-    <Compile Update="$(WindowsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+    <Compile
+        Condition=" '$(TargetPlatformIdentifier)' == 'windows' "
+        Update="$(WindowsProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
-  </ItemGroup>
-
-  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'tizen' ">
-    <Compile Update="$(TizenProjectFolder)**/*$(DefaultLanguageSourceExtension)">
+    <Compile
+        Condition=" '$(TargetPlatformIdentifier)' == 'tizen' "
+        Update="$(TizenProjectFolder)**/*$(DefaultLanguageSourceExtension)">
       <ExcludeFromCurrentConfiguration>false</ExcludeFromCurrentConfiguration>
     </Compile>
   </ItemGroup>

